src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,262 @@
+
+(defgeneric tref (tensor &rest pos)
+  (:documentation "Get a reference to element at position"))
+
+(defgeneric (setf tref) (new-value tensor &rest pos)
+  (:documentation "Set the value of element at position"))
+
+(defgeneric copy-tensor (tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, creating new storage without initializing contents"))
+
+(defgeneric deep-copy-tensor (tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, copying the contents of the storage"))
+
+;;; abstract-tensor generic methods
+;;; These methods have default generic implementations but can be optimized by subclasses
+
+(defgeneric order (tensor)
+  (:documentation "Order (number of dimensions) of the tensor")
+  (:method ((tensor abstract-tensor))
+    (length (shape tensor))))
+
+(defgeneric size (tensor)
+  (:documentation "Total number of elements in the tensor")
+  (:method ((tensor abstract-tensor))
+    (reduce #'* (shape tensor))))
+
+(defgeneric element-type (tensor)
+  (:documentation "The type of the elements in the tensor")
+  (:method ((tensor abstract-tensor))
+    (declare (ignore tensor))
+    t))
+
+(defgeneric lisp-array (tensor &optional target)
+  (:documentation "Return a lisp array containing the data from the tensor
+
+If TARGET is specified then the contents of the tensor are copied into the array.
+In the event TARGET is not specified, the result may return an array sharing memory with the input tensor.")
+  (:method :before (tensor &optional target)
+    (unless (null target)
+      (policy-cond:policy-if
+       (< speed safety)
+       (assert (and (= (order tensor) (array-rank target))
+                    (equal (shape tensor) (array-dimensions target))))
+       nil)))
+  (:method ((tensor abstract-tensor) &optional target)
+    (let  ((arr (or target (make-array (shape tensor) :element-type (element-type tensor)))))
+      (map-indexes
+       (shape tensor)
+       (lambda (&rest pos)
+         (let ((val (apply #'tref tensor pos)))
+           (apply #'(setf aref) val arr pos))))
+      arr)))
+
+(defgeneric map! (function tensor)
+  (:documentation "Map elements of TENSOR by replacing the value with the output of FUNCTION on the element")
+  (:method ((function function) (tensor abstract-tensor))
+    (map-indexes
+     (shape tensor)
+     (lambda (&rest dims)
+       (apply #'(setf tref) (funcall function (apply #'tref tensor dims)) tensor dims)))
+    tensor))
+
+(defgeneric into! (function tensor)
+  (:documentation "Map indices of TENSOR by replacing the value with the output of FUNCTION on the index at the element.")
+  (:method ((function function) (tensor abstract-tensor))
+    (map-indexes
+     (shape tensor)
+     (lambda (&rest dims)
+       (apply #'(setf tref) (apply function dims) tensor dims)))
+    tensor))
+
+(defgeneric foreach (function tensor)
```

Comment:
It might make more sense for this to be a macro but don't do it in this PR. We can write a `do-tensor` macro for instance to do the job later on.